packages/strapi-plugin-content-manager/controllers/collection-types.js

```diff
@@ -0,0 +1,226 @@
+
+  async findOne(ctx) {
+    const { userAbility } = ctx.state;
+    const { model, id } = ctx.params;
+
+    const entityManager = getService('entity-manager');
+    const permissionChecker = getService('permission-checker').create({ userAbility, model });
+
+    if (permissionChecker.cannot.read()) {
+      return ctx.forbidden();
+    }
+
+    const entity = await entityManager.findOneWithCreatorRoles(id, model);
+
+    if (!entity) {
+      return ctx.notFound();
+    }
+
+    if (permissionChecker.cannot.read(entity)) {
+      return ctx.forbidden();
+    }
+
+    ctx.body = permissionChecker.sanitizeOutput(entity);
+  },
+
+  async create(ctx) {
+    const { userAbility, user } = ctx.state;
+    const { model } = ctx.params;
+    const { body } = ctx.request;
+
+    const entityManager = getService('entity-manager');
+    const permissionChecker = getService('permission-checker').create({ userAbility, model });
+
+    if (permissionChecker.cannot.create()) {
+      return ctx.forbidden();
+    }
+
+    const pickWritables = pickWritableAttributes({ model });
+    const pickPermittedFields = permissionChecker.sanitizeCreateInput;
+    const setCreator = setCreatorFields({ user });
+
+    const sanitizeFn = pipe([pickWritables, pickPermittedFields, setCreator]);
+
+    await wrapBadRequest(async () => {
+      const entity = await entityManager.create(sanitizeFn(body), model);
+      ctx.body = permissionChecker.sanitizeOutput(entity);
+
+      await strapi.telemetry.send('didCreateFirstContentTypeEntry', { model });
+    })();
+  },
+
+  async update(ctx) {
+    const { userAbility, user } = ctx.state;
+    const { id, model } = ctx.params;
+    const { body } = ctx.request;
+
+    const entityManager = getService('entity-manager');
+    const permissionChecker = getService('permission-checker').create({ userAbility, model });
+
+    if (permissionChecker.cannot.update()) {
+      return ctx.forbidden();
+    }
+
+    const entity = await entityManager.findOneWithCreatorRoles(id, model);
+
+    if (!entity) {
+      return ctx.notFound();
+    }
+
+    if (permissionChecker.cannot.update(entity)) {
+      return ctx.forbidden();
+    }
+
+    const pickWritables = pickWritableAttributes({ model });
+    const pickPermittedFields = permissionChecker.sanitizeUpdateInput(entity);
+    const setCreator = setCreatorFields({ user, isEdition: true });
+
+    const sanitizeFn = pipe([pickWritables, pickPermittedFields, setCreator]);
+
+    await wrapBadRequest(async () => {
+      const updatedEntity = await entityManager.update(entity, sanitizeFn(body), model);
+
+      ctx.body = permissionChecker.sanitizeOutput(updatedEntity);
+    })();
+  },
+
+  async delete(ctx) {
+    const { userAbility } = ctx.state;
+    const { id, model } = ctx.params;
+
+    const entityManager = getService('entity-manager');
+    const permissionChecker = getService('permission-checker').create({ userAbility, model });
+
+    if (permissionChecker.cannot.delete()) {
```

Comment:
Yes they are. I can remove them but I link the policies for this are actually not good. It means the function can be called with another mean and skipp the verification. We are splitting the authorization logic into two different places and making it hard to evolve.

Other note. Since we have json routes we are using the full action name instead of constants for them which makes it even more fragile.

I can remove them but we definitely will need to collocate the permissin logic in the future. Supporting something like
the following or using decorators.

```js
module.exports = {
  find: [hasPermission(ACTIONS.read), ctx => { /*...*/}]
}
```